File: lp/comm/layer0/sym.go
package layer0

import (
	"../../../crypto/sym"
	"../handler"
)

type EncryptedHandler struct {
	key        []byte
	subhandler handler.Handler
}

func NewEncryptedHandler(key []byte, subhandler handler.Handler) *EncryptedHandler {
	return &EncryptedHandler{
		key:        key,
		subhandler: subhandler,
	}
}

func (h *EncryptedHandler) HandleMsg(msg []byte) ([]byte, error) {
	decryptedMsg, err := sym.ValidateThenDecryptMessage(msg, h.key)
	if err != nil {
		return nil, err
	}
	response, err := h.subhandler.HandleMsg(decryptedMsg)
	if err != nil {
		return nil, err
	}
	encryptedResponse, err := sym.EncryptThenSignMessage(response, h.key)
	if err != nil {
		return nil, err
	}
	return encryptedResponse, nil
}

File: python/malon_lp/malon_lp/listener/sym.py

from .handler import Handler

from malon_common.crypto.sym import SymmetricCipher

class EncryptedHandler:
    def __init__(self, key: bytes, handler: Handler):
        self._cipher = SymmetricCipher(key)
        self._handler = handler
    
    def handle_msg(self, msg: bytes) -> bytes:
        decrypted_msg = self._cipher.verify_decrypt_msg(msg)
        response_msg = self._handler.handle_msg(decrypted_msg)
        encrypted_response_msg = self._cipher.encrypt_sign_msg(response_msg)
        return encrypted_response_msg
    